compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorPad.cpp

```diff
@@ -0,0 +1,174 @@
+    PatternRewriter &rewriter, linalg::LinalgOp linalgOp,
+    linalg::LinalgOp &paddedOp) {
+  Location loc = linalgOp.getLoc();
+
+  PatternRewriter::InsertionGuard g(rewriter);
+  // Set IP after op because we also take the dims of the original output.
+  rewriter.setInsertionPointAfter(linalgOp);
+
+  // Pad each input operand in shared memory out to 32,32
+  SmallVector<int64_t> paddedShape = {32, 32};
+  SmallVector<Value> paddedOperands;
+  paddedOperands.reserve(linalgOp.getNumInputsAndOutputs());
+  for (OpOperand *opOperand : linalgOp.getInputAndOutputOperands()) {
+    // Find DispatchTensorLoadOp's feeding into the linalg or abort.
+    auto tensorLoad = dyn_cast_or_null<IREE::Flow::DispatchTensorLoadOp>(
+        opOperand->get().getDefiningOp());
+    if (!tensorLoad) {
+      return rewriter.notifyMatchFailure(linalgOp, "does not have tensor load");
+    }
+
+    Value paddingValue = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(
+                 tensorLoad.getType().cast<ShapedType>().getElementType()));
+    auto paddedTensorType0 =
+        RankedTensorType::get(paddedShape, getElementTypeOrSelf(tensorLoad));
+    Value paddedValue = linalg::makeComposedPadHighOp(
+        rewriter, loc, paddedTensorType0, tensorLoad, paddingValue, false);
+    paddedOperands.push_back(paddedValue);
+  }
+
+  // Clone linalgOp to paddedOp with padded input/output shapes.
+  auto resultTensorTypes =
+      ValueRange(paddedOperands).take_back(linalgOp.getNumOutputs()).getTypes();
+  paddedOp = linalgOp.clone(rewriter, loc, resultTensorTypes, paddedOperands);
+
+  // Slice out the original shape from the padded result to pass on to
+  // consumers. The original linalg op is used to provide the dims for the reify
+  // result shapes.
+  SmallVector<SmallVector<Value>> reifiedResultShapes;
+  if (failed(cast<ReifyRankedShapedTypeOpInterface>(linalgOp.getOperation())
+                 .reifyResultShapes(rewriter, reifiedResultShapes))) {
+    return failure();
+  }
+
+  SmallVector<Value> paddedSubviewResults;
+  paddedSubviewResults.reserve(paddedOp->getNumResults());
+  for (const auto &en : llvm::enumerate(paddedOp->getResults())) {
+    Value paddedResult = en.value();
+    int64_t resultNumber = en.index();
+    int64_t rank = paddedResult.getType().cast<RankedTensorType>().getRank();
+    SmallVector<OpFoldResult> offsets(rank, rewriter.getIndexAttr(0));
+    SmallVector<OpFoldResult> sizes;
+    for (Value v : reifiedResultShapes[resultNumber])
+      sizes.push_back(getAsOpFoldResult(v));
+    SmallVector<OpFoldResult> strides(rank, rewriter.getIndexAttr(1));
+    paddedSubviewResults.push_back(rewriter.create<tensor::ExtractSliceOp>(
+        loc, paddedResult, offsets, sizes, strides));
+  }
+  return paddedSubviewResults;
+}
+
+static bool hasTwoLoops(linalg::LinalgOp linalgOp) {
+  return linalgOp.getNumParallelLoops() == 2;
+}
+
+struct TransposePadOpPattern : public OpRewritePattern<linalg::GenericOp> {
```

Comment:
You don't really need a pattern here, I know we have been heavily relying on rewrite patterns combined with attributes but we are moving away from them and moving to applying direct transformation using APIs as you have above `FailureOr<SmallVector<Value>> rewriteAsPaddedOp`. This will ensure easier transition to transform dialect.
You should be able to simplify this code and remove the need for an attribute this way. You can just walk the function and apply the `rewriteAsPaddedOp` based on the precondition.

Reply:
Using Rewriter is fine, and it can be done without using `OpRewritePattern`. The pattern rewrite will run to fix point which is really not what we want in this case so not using it should make a lot of things simpler.



> @rsuderman also advised against directly manipulating IR with complicated changes as unintended side effects persist if the pass fails/bails out early. I do prefer the error handling of the rewriter (its a nice bonus).

The bailing out of the rewrite pattern really doesn't buy you much as it doesn't clean up anything. You still need to first do all the precondition checks then do the transformation.

> I can't speak to the Transform Dialect ( I don't have any details on it). Can I get more context here? (we can discuss in person).

Yes that probably requires a long answer.